pkg/cmd/pulumi/convert-trace.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/google/pprof/profile"
+	"github.com/spf13/cobra"
+	"sourcegraph.com/sourcegraph/appdash"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+// each root is sampled at a particular granularity. the nested traces are converted into a callstack at each
+// sampling point.
+
+func traceRoots(querier appdash.Queryer) ([]*appdash.Trace, error) {
+	traces, err := querier.Traces(appdash.TracesOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	var roots []*appdash.Trace
+	for _, t := range traces {
+		if t.ID.Parent == 0 {
+			roots = append(roots, t)
+		}
+	}
+
+	return roots, nil
+}
+
+type traceSample struct {
+	when  time.Duration
+	where []string
+}
+
+func (s *traceSample) key() string {
+	return strings.Join(s.where, "<")
+}
+
+func findGRPCPayload(t *appdash.Trace) string {
+	for _, a := range t.Annotations {
+		if a.Key == "Msg" {
+			var msg map[string]interface{}
+			if err := json.Unmarshal(a.Value, &msg); err == nil {
+				if req, ok := msg["gRPC request"]; ok {
+					s, _ := req.(string)
+					return s
+				}
+			}
+		}
+	}
+	return ""
+}
+
+var urnFieldRe = regexp.MustCompile(`urn:"([^"]+)"`)
+var tokFieldRe = regexp.MustCompile(`tok:"([^"]+)"`)
+
+func extractResourceType(t *appdash.Trace) string {
+	grpcPayload := findGRPCPayload(t)
+	matches := urnFieldRe.FindStringSubmatch(grpcPayload)
+	if len(matches) == 0 {
+		return ""
+	}
+	urn := resource.URN(matches[1])
```

Comment:
I believe `FindStringSubmatch` will always return 0 or 2 results for the current regexp, but it would probably be good to check for `len(matches) < 2` just to be safe.

Reply:
This will always return 0 or 2 results--I don't think that we need to sanity check.